File: langs/dotnet/src/Dotlyte/Encryption.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace Dotlyte;

/// <summary>
/// AES-256-GCM encryption/decryption for DOTLYTE v2 (SOPS-style).
/// Format: ENC[aes-256-gcm,iv:base64,data:base64,tag:base64]
/// </summary>
public static class Encryption
{
    private const string Prefix = "ENC[";
    private const int KeyBytes = 32;
    private const int IvBytes = 12;
    private const int TagBytes = 16;

    /// <summary>
    /// Check whether a value is an encrypted SOPS-style string.
    /// </summary>
    public static bool IsEncrypted(object? value)
    {
        return value is string s
            && s.StartsWith(Prefix, StringComparison.Ordinal)
            && s.EndsWith(']');
    }

    /// <summary>
    /// Generate a random 32-byte key as a hex string.
    /// </summary>
    public static string GenerateKey()
    {
        var bytes = RandomNumberGenerator.GetBytes(KeyBytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    /// <summary>
    /// Encrypt a plaintext string with a hex-encoded 256-bit key.
    /// </summary>
    /// <exception cref="DotlyteException">Thrown on encryption failure.</exception>
    public static string EncryptValue(string plaintext, string keyHex)
    {
        var keyBytes = Convert.FromHexString(keyHex);
        if (keyBytes.Length != KeyBytes)
            throw new DotlyteException("Encryption key must be 32 bytes (64 hex chars)");

        var iv = RandomNumberGenerator.GetBytes(IvBytes);
        var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
        var ciphertext = new byte[plaintextBytes.Length];
        var tag = new byte[TagBytes];

        using var aes = new AesGcm(keyBytes, TagBytes);
        aes.Encrypt(iv, plaintextBytes, ciphertext, tag);

        var ivB64 = Convert.ToBase64String(iv);
        var dataB64 = Convert.ToBase64String(ciphertext);
        var tagB64 = Convert.ToBase64String(tag);

        return $"ENC[aes-256-gcm,iv:{ivB64},data:{dataB64},tag:{tagB64}]";
    }

    /// <summary>
    /// Decrypt a SOPS-style encrypted value.
    /// </summary>
    /// <exception cref="DecryptionException">Thrown on decryption failure.</exception>
    public static string DecryptValue(string encrypted, string keyHex)
    {
        var keyBytes = Convert.FromHexString(keyHex);
        if (keyBytes.Length != KeyBytes)
            throw new DecryptionException("Encryption key must be 32 bytes (64 hex chars)");

        var inner = encrypted;
        if (inner.StartsWith("ENC[", StringComparison.Ordinal) && inner.EndsWith(']'))
            inner = inner[4..^1];

        string? ivB64 = null, dataB64 = null, tagB64 = null;
        foreach (var part in inner.Split(','))
        {
            var trimmed = part.Trim();
            if (trimmed.StartsWith("iv:", StringComparison.Ordinal))
                ivB64 = trimmed[3..];
            else if (trimmed.StartsWith("data:", StringComparison.Ordinal))
                dataB64 = trimmed[5..];
            else if (trimmed.StartsWith("tag:", StringComparison.Ordinal))
                tagB64 = trimmed[4..];
        }

        if (ivB64 is null || dataB64 is null || tagB64 is null)
            throw new DecryptionException("Decryption failed: malformed ENC[] value");

        try
        {
            var iv = Convert.FromBase64String(ivB64);
            var ciphertext = Convert.FromBase64String(dataB64);
            var tag = Convert.FromBase64String(tagB64);
            var plaintext = new byte[ciphertext.Length];

            using var aes = new AesGcm(keyBytes, TagBytes);
            aes.Decrypt(iv, ciphertext, tag, plaintext);

            return Encoding.UTF8.GetString(plaintext);
        }
        catch (CryptographicException ex)
        {
            throw new DecryptionException(
                "Decryption failed: invalid key or corrupted data", null, ex);
        }
    }

    /// <summary>
    /// Decrypt all ENC[...] values in a dictionary (recursive).
    /// </summary>
    public static Dictionary<string, object?> DecryptDictionary(
        Dictionary<string, object?> data, string keyHex)
    {
        var result = new Dictionary<string, object?>();
        foreach (var (k, v) in data)
        {
            if (v is Dictionary<string, object?> nested)
                result[k] = DecryptDictionary(nested, keyHex);
            else if (IsEncrypted(v))
                result[k] = Coercion.Coerce(DecryptValue((string)v!, keyHex));
            else
                result[k] = v;
        }
        return result;
    }

    /// <summary>
    /// Resolve encryption key from environment / key file.
    /// </summary>
    public static string? ResolveEncryptionKey(string? envName = null)
    {
        if (!string.IsNullOrEmpty(envName))
        {
            var envSpecific = Environment.GetEnvironmentVariable(
                $"DOTLYTE_KEY_{envName.ToUpperInvariant()}");
            if (!string.IsNullOrEmpty(envSpecific))
                return envSpecific;
        }

        var globalKey = Environment.GetEnvironmentVariable("DOTLYTE_KEY");
        if (!string.IsNullOrEmpty(globalKey))
            return globalKey;

        const string keyFile = ".dotlyte-keys";
        if (File.Exists(keyFile))
        {
            var lines = File.ReadAllLines(keyFile);
            var first = lines.FirstOrDefault(l => !string.IsNullOrWhiteSpace(l));
            if (!string.IsNullOrEmpty(first))
                return first.Trim();
        }

        return null;
    }

    /// <summary>
    /// Re-encrypt all values from an old key to a new key.
    /// Only values matching the ENC[...] format are re-encrypted.
    /// </summary>
    /// <param name="data">Dictionary of potentially encrypted string values.</param>
    /// <param name="oldKey">The current encryption key (hex).</param>
    /// <param name="newKey">The new encryption key (hex).</param>
    /// <returns>A new dictionary with values re-encrypted under the new key.</returns>
    /// <exception cref="DotlyteException">Thrown on decryption or encryption failure.</exception>
    public static Dictionary<string, string> RotateKeys(
        Dictionary<string, string> data, string oldKey, string newKey)
    {
        var result = new Dictionary<string, string>();
        foreach (var (k, v) in data)
        {
            if (IsEncrypted(v))
            {
                var plaintext = DecryptValue(v, oldKey);
                result[k] = EncryptValue(plaintext, newKey);
            }
            else
            {
                result[k] = v;
            }
        }
        return result;
    }

    /// <summary>
    /// Attempt to decrypt a value using a list of candidate keys.
    /// Returns the decrypted value from the first key that succeeds, or <c>null</c> if none work.
    /// </summary>
    /// <param name="keys">Array of hex-encoded encryption keys to try.</param>
    /// <param name="encryptedValue">The ENC[...] encrypted value.</param>
    /// <returns>The decrypted plaintext, or <c>null</c> if no key succeeded.</returns>
    public static string? ResolveKeyWithFallback(string[] keys, string encryptedValue)
    {
        foreach (var key in keys)
        {
            try
            {
                return DecryptValue(encryptedValue, key);
            }
            catch (DotlyteException)
            {
                // Try next key
            }
        }
        return null;
    }

    /// <summary>
    /// Encrypt all values in a dictionary, optionally limited to sensitive keys only.
    /// </summary>
    /// <param name="data">The plaintext key-value pairs.</param>
    /// <param name="key">The hex-encoded encryption key.</param>
    /// <param name="sensitiveKeys">If provided, only these keys are encrypted; others pass through.</param>
    /// <returns>A new dictionary with encrypted values.</returns>
    public static Dictionary<string, string> EncryptVault(
        Dictionary<string, string> data, string key, HashSet<string>? sensitiveKeys = null)
    {
        var result = new Dictionary<string, string>();
        foreach (var (k, v) in data)
        {
            if (sensitiveKeys is null || sensitiveKeys.Contains(k))
            {
                result[k] = EncryptValue(v, key);
            }
            else
            {
                result[k] = v;
            }
        }
        return result;
    }

    /// <summary>
    /// Decrypt all ENC[...] values in a string dictionary.
    /// Non-encrypted values pass through unchanged.
    /// </summary>
    /// <param name="data">The dictionary with potentially encrypted values.</param>
    /// <param name="key">The hex-encoded encryption key.</param>
    /// <returns>A new dictionary with all values decrypted.</returns>
    /// <exception cref="DecryptionException">Thrown when a value cannot be decrypted.</exception>
    public static Dictionary<string, string> DecryptVault(
        Dictionary<string, string> data, string key)
    {
        var result = new Dictionary<string, string>();
        foreach (var (k, v) in data)
        {
            if (IsEncrypted(v))
            {
                result[k] = DecryptValue(v, key);
            }
            else
            {
                result[k] = v;
            }
        }
        return result;
    }
}

File: langs/ruby/lib/dotlyte/encryption.rb
# frozen_string_literal: true

require "openssl"
require "base64"
require "securerandom"

module Dotlyte
  # AES-256-GCM encryption/decryption for DOTLYTE v2 (SOPS-style).
  #
  # Format: ENC[aes-256-gcm,iv:<base64>,data:<base64>,tag:<base64>]
  module Encryption
    PREFIX = "ENC["
    GCM_TAG_BYTES = 16
    IV_BYTES = 12
    KEY_BYTES = 32

    # Check whether a string is an encrypted SOPS-style value.
    def self.encrypted?(value)
      value.is_a?(String) && value.start_with?(PREFIX) && value.end_with?("]")
    end

    # Generate a random 32-byte key as hex.
    def self.generate_key
      SecureRandom.hex(KEY_BYTES)
    end

    # Encrypt a plaintext string with a hex-encoded 256-bit key.
    #
    # @param plaintext [String]
    # @param key_hex [String] 64-char hex string
    # @return [String] SOPS-style encrypted value
    def self.encrypt_value(plaintext, key_hex)
      key_bytes = [key_hex].pack("H*")
      raise Error, "Key must be 32 bytes, got #{key_bytes.bytesize}" unless key_bytes.bytesize == KEY_BYTES

      cipher = OpenSSL::Cipher.new("aes-256-gcm")
      cipher.encrypt
      cipher.key = key_bytes
      iv = cipher.random_iv
      cipher.auth_data = ""

      ciphertext = cipher.update(plaintext) + cipher.final
      tag = cipher.auth_tag(GCM_TAG_BYTES)

      iv_b64 = Base64.strict_encode64(iv)
      data_b64 = Base64.strict_encode64(ciphertext)
      tag_b64 = Base64.strict_encode64(tag)

      "ENC[aes-256-gcm,iv:#{iv_b64},data:#{data_b64},tag:#{tag_b64}]"
    rescue OpenSSL::Cipher::CipherError => e
      raise Error, "Encryption failed: #{e.message}"
    end

    # Decrypt a SOPS-style encrypted value.
    #
    # @param encrypted [String]
    # @param key_hex [String] 64-char hex string
    # @return [String] plaintext
    def self.decrypt_value(encrypted, key_hex)
      key_bytes = [key_hex].pack("H*")
      raise Error, "Key must be 32 bytes, got #{key_bytes.bytesize}" unless key_bytes.bytesize == KEY_BYTES

      inner = encrypted
      inner = inner[4..-2] if inner.start_with?("ENC[") && inner.end_with?("]")

      parts = {}
      inner.split(",").each do |part|
        part = part.strip
        if part.start_with?("iv:")
          parts[:iv] = part[3..]
        elsif part.start_with?("data:")
          parts[:data] = part[5..]
        elsif part.start_with?("tag:")
          parts[:tag] = part[4..]
        end
      end

      iv = Base64.strict_decode64(parts[:iv])
      data = Base64.strict_decode64(parts[:data])
      tag = Base64.strict_decode64(parts[:tag])

      cipher = OpenSSL::Cipher.new("aes-256-gcm")
      cipher.decrypt
      cipher.key = key_bytes
      cipher.iv = iv
      cipher.auth_tag = tag
      cipher.auth_data = ""

      cipher.update(data) + cipher.final
    rescue OpenSSL::Cipher::CipherError => e
      raise DecryptionError, "Decryption failed: #{e.message}"
    end

    # Decrypt all ENC[...] values in a hash.
    def self.decrypt_hash(data, key_hex)
      data.each do |k, v|
        data[k] = decrypt_value(v, key_hex) if encrypted?(v)
      end
    end

    # Resolve encryption key from environment / key file.
    def self.resolve_encryption_key(env_name = nil)
      if env_name && !env_name.empty?
        val = ENV["DOTLYTE_KEY_#{env_name.upcase}"]
        return val if val && !val.empty?
      end

      val = ENV["DOTLYTE_KEY"]
      return val if val && !val.empty?

      key_file = ".dotlyte-keys"
      if File.exist?(key_file)
        line = File.readlines(key_file, chomp: true).first&.strip
        return line if line && !line.empty?
      end

      nil
    end

    # Re-encrypt all encrypted values in a hash with a new key.
    #
    # @param data [Hash] hash potentially containing ENC[...] values
    # @param old_key [String] current encryption key (hex)
    # @param new_key [String] new encryption key (hex)
    # @return [Hash] hash with values re-encrypted under new_key
    # @raise [Dotlyte::Error] if decryption with old_key fails
    def self.rotate_keys(data, old_key:, new_key:)
      data.each_with_object({}) do |(k, v), result|
        result[k] = if v.is_a?(Hash)
                      rotate_keys(v, old_key: old_key, new_key: new_key)
                    elsif encrypted?(v)
                      plaintext = decrypt_value(v, old_key)
                      encrypt_value(plaintext, new_key)
                    else
                      v
                    end
      end
    end

    # Try each key in order to decrypt a value. Returns the first working key or nil.
    #
    # @param keys [Array<String>] candidate encryption keys (hex)
    # @param encrypted_value [String] the encrypted value to test
    # @return [String, nil] the working key, or nil if none decrypt successfully
    def self.resolve_key_with_fallback(keys, encrypted_value)
      keys.each do |candidate|
        begin
          decrypt_value(encrypted_value, candidate)
          return candidate
        rescue DecryptionError, Error
          next
        end
      end
      nil
    end

    # Encrypt selected values in a hash (vault-style).
    #
    # @param data [Hash] plaintext hash
    # @param key [String] encryption key (hex)
    # @param sensitive_keys [Array<String>, Set<String>, nil] keys to encrypt (nil = all)
    # @return [Hash] hash with encrypted values
    def self.encrypt_vault(data, key:, sensitive_keys: nil)
      targets = sensitive_keys ? Set.new(sensitive_keys) : nil

      data.each_with_object({}) do |(k, v), result|
        result[k] = if v.is_a?(Hash)
                      encrypt_vault(v, key: key, sensitive_keys: sensitive_keys)
                    elsif v.is_a?(String) && !encrypted?(v) && (targets.nil? || targets.include?(k))
                      encrypt_value(v, key)
                    else
                      v
                    end
      end
    end

    # Decrypt all encrypted values in a hash (vault-style).
    #
    # @param data [Hash] hash with ENC[...] values
    # @param key [String] encryption key (hex)
    # @return [Hash] hash with all values decrypted
    def self.decrypt_vault(data, key:)
      data.each_with_object({}) do |(k, v), result|
        result[k] = if v.is_a?(Hash)
                      decrypt_vault(v, key: key)
                    elsif encrypted?(v)
                      decrypt_value(v, key)
                    else
                      v
                    end
      end
    end
  end
end
